### PR TITLE
fix(experimental elaborator): Fix duplicate `resolve_type` on self type and don't leak a trait impl's generics

### PR DESCRIPTION
**Pull request by jfecher**
# Description

## Problem\*

Resolves 

## Summary\*

Fixes two issues:
- Use cached `resolved_object_type` instead of re-resolving the self type in `declare_methods_on_struct` and `check_trait_impl_coherence`. This is needed to prevent duplicate errors, and in the later case, since generics are not added to scope to be able to re-resolve the object type at that point.
- Fix a case when resolving trait impls that `self.generics` was not cleared afterward.

## Additional Context

Down to 112 errors in the standard library now. The remaining errors are variants of:
- "Duplicate definitions of \<global> found"
- "Duplicate definitions of \<generic> found" - looks to be only one case where there is a generic directly on a trait method. `fn hash<H>(self, state: &mut H) where H: Hasher;`
- "Expression type is ambiguous" - from the two turbofish cases that haven't been merged in this branch yet
- "Expected type &mut _, found type H" - issues with auto-deref and trait methods possibly
- "No matching impl found for ..." - trait solver issues

## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by jfecher** [comment]
To avoid fixing as many merge conflicts I'm going to keep this PR open and merge jf/elaborator-fixes5 through 8 into this one before merging this into master

File: compiler/noirc_frontend/src/elaborator/mod.rs [hunk 1]
```diff
@@ -1306,8 +1313,10 @@ impl<'context> Elaborator<'context> {
             let unresolved_type = &trait_impl.object_type;
             let self_type_span = unresolved_type.span;
             self.add_generics(&trait_impl.generics);
+            trait_impl.resolved_generics = self.generics.clone();
```
Comment (vezenovm) [inline compiler/noirc_frontend/src/elaborator/mod.rs:1316]: I mentioned this in the other PR review, but`recover_generics` could be used here?

**Review by vezenovm — APPROVED** [review]

File: compiler/noirc_frontend/src/elaborator/mod.rs (hunk 1 above)
Comment (jfecher) [reply compiler/noirc_frontend/src/elaborator/mod.rs:1316]: This case is simpler with just `generics.clear()` I think but it'd be the same either way

**Comment by jfecher** [comment]
The diff in fixes7 got messed up so I'm going to merge master into both of these PRs anyway and merge this one to master.

**Pull request merged**